### PR TITLE
[3882] Add a new tree event to support non explorer tree descriptions

### LINKED ISSUE #3882: Add a new tree event to handle tree description which are not explorer-related
* [x] **I have checked that this feature has not yet been suggested by someone else.**

At the moment, sirius web only provides TreeDescription for explorers.
First, we refactored existing code to reflect this reality (issue https://github.com/eclipse-sirius/sirius-web/issues/3875).

Now, in this issue, we introduce a new tree event that allows us to handle tree descriptions that are not explorer-related.

### PR DESCRIPTION
**Pull request by jerome-obeo**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/trees/frontend/sirius-components-trees/src/trees/TreeRepresentation.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  const onExpandedElementChange = (expanded: string[], maxDepth: number) => {
+    setState((prevState) => ({ ...prevState, expanded, maxDepth }));
+  };
+
+  return (
+    <div>
+      <TreeView
+        editingContextId={editingContextId}
+        readOnly={false}
```
Comment (frouene) [inline packages/trees/frontend/sirius-components-trees/src/trees/TreeRepresentation.tsx]: You should not hardcode read-only to false, but use the props readOnly from the `RepresentationComponentProps`

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IInput input) {
+        this.counter.increment();
+
+        String message = this.messageService.invalidInput(input.getClass().getSimpleName(), CreateRepresentationInput.class.getSimpleName());
+        IPayload payload = new ErrorPayload(input.id(), message);
+        ChangeDescription changeDescription = new ChangeDescription(ChangeKind.NOTHING, editingContext.getId(), input);
+
+        if (input instanceof CreateRepresentationInput createRepresentationInput) {
+            Optional<TreeDescription> optionalDeckDescription = this.representationDescriptionSearchService.findById(editingContext, createRepresentationInput.representationDescriptionId())
+                    .filter(TreeDescription.class::isInstance)
```
Comment (frouene) [inline packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java:96]: should be renamed to optionalTreeDescription

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                    .map(TreeDescription.class::cast);
+            Optional<Object> optionalObject = this.objectService.getObject(editingContext, createRepresentationInput.objectId());
+
+            if (optionalDeckDescription.isPresent() && optionalObject.isPresent()) {
+                TreeDescription deckDescription = optionalDeckDescription.get();
+                Object object = optionalObject.get();
```
Comment (frouene) [inline packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java:102]: should be renamed to treeDescription

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/api/TreeCreationParameters.java [hunk 4]
```diff
@@ -37,6 +37,8 @@ public final class TreeCreationParameters {
 
     private IEditingContext editingContext;
 
+    private Object object;
+
```
Comment (frouene) [inline packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/api/TreeCreationParameters.java:41]: I find the term object a little too generic, perhaps targetObject would be a little better.

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/TreeNonSemanticElementControllerTests.java [hunk 5]
```diff
@@ -45,21 +46,29 @@
 import org.springframework.transaction.annotation.Transactional;
 
 import graphql.execution.DataFetcherResult;
+import reactor.core.publisher.Flux;
 import reactor.test.StepVerifier;
 
 /**
- * Integration tests of a non semantic element in the explorer.
+ * Integration tests of a non semantic element in the .
  *
```
Comment (frouene) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/TreeNonSemanticElementControllerTests.java:55]: in the ??

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/trees/frontend/sirius-components-trees/src/trees/TreeRepresentation.tsx (hunk 1 above)
Comment (jerome-obeo) [reply packages/trees/frontend/sirius-components-trees/src/trees/TreeRepresentation.tsx]: Done.

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java (hunk 2 above)
Comment (jerome-obeo) [reply packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java:96]: Done

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java (hunk 3 above)
Comment (jerome-obeo) [reply packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/handlers/CreateTreeEventHandler.java:102]: Done

File: packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/api/TreeCreationParameters.java (hunk 4 above)
Comment (jerome-obeo) [reply packages/trees/backend/sirius-components-collaborative-trees/src/main/java/org/eclipse/sirius/components/collaborative/trees/api/TreeCreationParameters.java:41]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/TreeNonSemanticElementControllerTests.java (hunk 5 above)
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/TreeNonSemanticElementControllerTests.java:55]: Done

**Review by frouene — APPROVED** [review]

**Review by sbegaudeau — DISMISSED** [review]
I'll just block the merge of this feature for now since there are two PRs that are a bit related to this that I need to see merged first:

- [3763] Implements the Selection Dialog TreeView
- [3976] Remove IRepresentationConfiguration
- And then this one

Since both PR already had to experience some major impact due to the integration of the previous PR on this topic, I have to make sure they get merged now that they just finish rebasing their work.

**Comment by frouene** [comment]
> I'll just block the merge of this feature for now since there are two PRs that are a bit related to this that I need to see merged first:
> 
>     * [3763] Implements the Selection Dialog TreeView
> 
>     * [3976] Remove IRepresentationConfiguration
> 
>     * And then this one
> 
> 
> Since both PR already had to experience some major impact due to the integration of the previous PR on this topic, I have to make sure they get merged now that they just finish rebasing their work.

Could be merge now that the other pending PRs have been integrated

**Pull request merged**